File: GlazeWM.Domain/Common/Enums/TilingDirection.cs
using System;

namespace GlazeWM.Domain.Common.Enums
{
  public enum TilingDirection
  {
    Vertical,
    Horizontal,
  }

  public static class TilingDirectionExtensions
  {
    /// <summary>
    /// Get the inverse of a given tiling direction.
    /// </summary>
    /// <exception cref="ArgumentOutOfRangeException"></exception>
    public static TilingDirection Inverse(this TilingDirection tilingDirection)
    {
      return tilingDirection switch
      {
        TilingDirection.Vertical => TilingDirection.Horizontal,
        TilingDirection.Horizontal => TilingDirection.Vertical,
        _ => throw new ArgumentOutOfRangeException(nameof(tilingDirection)),
      };
    }
  }
}

File: packages/wm/src/common/tiling_direction.rs
use std::str::FromStr;

use anyhow::bail;
use serde::Serialize;

use super::Direction;

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(untagged, rename_all = "snake_case")]
pub enum TilingDirection {
  Vertical,
  Horizontal,
}

impl TilingDirection {
  /// Gets the inverse of a given tiling direction.
  ///
  /// Example:
  /// ```
  /// TilingDirection::Horizontal.inverse() // TilingDirection::Vertical
  /// ```
  pub fn inverse(&self) -> TilingDirection {
    match self {
      TilingDirection::Horizontal => TilingDirection::Vertical,
      TilingDirection::Vertical => TilingDirection::Horizontal,
    }
  }

  /// Gets the tiling direction that is needed when moving or shifting
  /// focus in a given direction.
  ///
  /// Example:
  /// ```
  /// TilingDirection::from_direction(Direction::Left) // TilingDirection::Horizontal
  /// ```
  pub fn from_direction(direction: &Direction) -> TilingDirection {
    match direction {
      Direction::Left | Direction::Right => TilingDirection::Horizontal,
      Direction::Up | Direction::Down => TilingDirection::Vertical,
    }
  }
}

impl FromStr for TilingDirection {
  type Err = anyhow::Error;

  /// Parses a string into a tiling direction.
  ///
  /// Example:
  /// ```
  /// TilingDirection::from_str("horizontal") // TilingDirection::Horizontal
  /// ```
  fn from_str(unparsed: &str) -> anyhow::Result<Self> {
    match unparsed {
      "horizontal" => Ok(TilingDirection::Horizontal),
      "vertical" => Ok(TilingDirection::Vertical),
      _ => bail!("Not a valid tiling direction: {}", unparsed),
    }
  }
}
